### ISSUE TITLE
Generate Sales Invoice Line Items from Timesheets

### ISSUE THREAD

**Issue by barredterra** [frappe/erpnext#26794]
Time tracking and Invoicing have been somewhat disconnected until now. We can pull **Timesheets** into a **Sales Invoice** but still have to add the corresponding line items manually.

## Proposed solution

I propose to allow mapping **Item** to **Activity Type** and fill the line items in **Sales Invoice** automatically, wherever possible.

- Add a new field to **Activity Type**, linking to the corresponding **Item**
- When **Timesheets** are fetched into a **Sales Invoice**, look up the linked **Items**. Add line items corresponding to the **Timesheets** and **Activity Types** (hours as quantity, rate from the timesheet).
- Provide an option via **Selling Settings** to sum similar **Timesheets** (same rate and Item) into one line item.

    Check interference with "Allow Item to be added multiple times in a transaction". Maybe we can just use this setting.

**Comment by anupamvs** [comment]
@barredterra We have discussed this internally. And it doesn't make sense to link items with Activity Type. You can attach the timesheet along with the Sales Invoice.

**Comment by barredterra** [comment]
@anupamvs can you please elaborate on why you think it doesn't make sense?

Do you have any other suggestions on how to automatically create sales invoice items from timesheets?

**Comment by rasos** [comment]
Agree that we should have an option to attach the time sheet entries to a sales invoice without the need to create a report for all timesheets from several employees. The table should include also the activity as set in the time sheet detail (instead of the empty Description field below). 

![TimeSheetDsecription](https://user-images.githubusercontent.com/485723/146642475-dacde3c2-3b30-496f-a8b3-f89d75e365c0.png)

**Comment by sagarvora** [comment]
This is an opt-in feature, people who do not want to automatically add items can just not map an activity type to an Item.

Proposed Implementation:
- Standardize Sales Invoice creation by creating a `make_sales_invoice` method that works like any other method used in **Create** or **Get Items From**.
- Allow filtering of Time Sheets by Customer, Project, From Time, To Time.
- Auto-create Sales Invoice Item(s) if Activity Type has been linked to an Item. The row's quantity will be the total of billing hours, and the row's amount will be the total billing amount corresponding to that activity type. Rate will be auto-calculated based on these values. We can make setting qty and rate automatically configurable using a settings DocType (Projects Settings / Selling Settings).
- (Optional) Auto-generate a description for the Item.

**Comment by barredterra** [comment]
@sagarvora I'm not sure about the rate. The alternative would be to set the quantity only and leave the rate as per price list / pricing rule. Maybe we can even make this configurable. What do you think?

**Comment by barredterra** [comment]
@rasos, not sure if we are talking about the same thing.

**Comment by rasos** [comment]
@barredterra - after reading again the proposal I understand that `adding line items` is referring here to the item list, and not to the list of time sheet entries which miss the description (and also the task) text. Probably I should open a separate issue on that.

**Comment by DrZoidberg09** [comment]
I am also highly interested in this. Are there any news?

**Comment by maveezmh** [comment]
Any updates on this???

**Comment by barredterra** [comment]
No updates, but I still think this would be a nice feature.

**Comment by jamesnsecond** [comment]
Thanks for the feature, it's great.
Would it be better if the Item code could be set in the Activity Type, not the popup for Get items from Timesheet?
Each Activity type needs to be billed at a different rate.

Our use case:
Timesheet1 - Activity Type: Labourer - Billing rate 1000
Timesheet2 - Activity Type: Electrician - Billing rate 1500
Timesheet3 - Activity Type: Electrician Supervisor - Billing rate 3000
Timesheet4 - Activity Type: Engineer - Billing rate 9000
Timesheet5 - Activity Type: Labourer - Billing rate 1000
Timesheet6 - Activity Type: Electrician - Billing rate 1500

Item Code: Labourer - Item Price 1000
Item Code: Electrician - Item Price 1500
Item Code: Electrician Supervisor - Item Price 3000
Item Code: Engineer - Item Price 3000

All of these activities are being recorded in timesheets for the same Project.
Ideally:
In a Sales Invoice, [Get items from Timesheets], select dates and project, do not select the item code, [Get timesheets].
Then the total number of hours for each activity type is totalled and billed as the item code set in the activity type.

Can we look at implementing @sagarvora 's suggestion?
https://github.com/frappe/erpnext/issues/26794#issuecomment-997211964

**Comment by maveezmh** [comment]
@barredterra instead of popup to select item code what if we define a service item in each of activity type.. the fetching of timesheet data will be much easier and clearly defined.

above PR will give option to select an item forwhole timesheet while the option i suggested willcreate invoice directly on single click

**Comment by barredterra** [comment]
@maveezmh did you read the discussion above? I've originally proposed this solution but it was rejected.

**Comment by williamluke4** [comment]
@anupamvs Do you remember why this was not considered a good idea?

**Comment by barredterra** [comment]
If we can collect enough 👍🏼 on this issue and someone is willing to fund development, I think we could still get the original proposal implemented.

**Comment by williamluke4** [comment]
I started some work on this here https://github.com/frappe/erpnext/pull/48707

**Comment by jamesnsecond** [comment]
I've noticed there is Another bug in this Timesheet to Sales Invoice functionality, but it was closed prematurely.

https://github.com/frappe/erpnext/issues/44167
https://github.com/frappe/erpnext/pull/44175

**Comment by komali2** [comment]
For our engineering co-op, this feature would be extremely helpful. I basically would like a way to generate a client invoice for a single project from all timesheets for that project, but grouped as invoice line items, as "Activities" or"Items" I don't particularly care, so long as the engineer's specific rate for that Activity is pulled in. This would dramatically improve our invoicing flow, to where I'm on upwork looking for someone to build this for us as a custom app. 

@barredterra do you have an estimate on what it would cost to fund this? I'm paying for it anyway for our own app as an override to `get_items_from_timesheets` most likely. 

Others, if you have similar flows, e.g. as @jamesnsecond , how are you managing it right now?

**Comment by komali2** [comment]
> I started some work on this here [#48707](https://github.com/frappe/erpnext/pull/48707)

Oh, just scanning through this PR, it looks like it on the right path, though it seems to be auto-mapping by name which is perhaps not so intuitive (the full solution would probably be a field on an activity explicitly mapping it to some item, I would expect?) but regardless, I'm curious why it was closed? 

Edit: I've hired somebody to implement this functionality as a custom app for our usecase. When they're done, I'll drop the code here.

**Comment by komali2** [comment]
> > I started some work on this here [#48707](https://github.com/frappe/erpnext/pull/48707)
> 
> Oh, just scanning through this PR, it looks like it on the right path, though it seems to be auto-mapping by name which is perhaps not so intuitive (the full solution would probably be a field on an activity explicitly mapping it to some item, I would expect?) but regardless, I'm curious why it was closed?
> 
> Edit: I've hired somebody to implement this functionality as a custom app for our usecase. When they're done, I'll drop the code here.

If you're interested, you can add this app to generate itemized line items per timesheet rates on sales invoices https://github.com/508-dev/timesheet_extended/tree/develope

**Comment by jamesnsecond** [comment]
> > > I started some work on this here [#48707](https://github.com/frappe/erpnext/pull/48707)
> > 
> > 
> > Oh, just scanning through this PR, it looks like it on the right path, though it seems to be auto-mapping by name which is perhaps not so intuitive (the full solution would probably be a field on an activity explicitly mapping it to some item, I would expect?) but regardless, I'm curious why it was closed?
> > Edit: I've hired somebody to implement this functionality as a custom app for our usecase. When they're done, I'll drop the code here.
> 
> If you're interested, you can add this app to generate itemized line items per timesheet rates on sales invoices https://github.com/508-dev/timesheet_extended/tree/develope

komali2, your contribution is greatly appreciated . We will endeavor to test it out as soon as possible and give some feedback.  Is it currently working well for your use case?

**Comment by komali2** [comment]
> > > > I started some work on this here [#48707](https://github.com/frappe/erpnext/pull/48707)
> > > 
> > > 
> > > Oh, just scanning through this PR, it looks like it on the right path, though it seems to be auto-mapping by name which is perhaps not so intuitive (the full solution would probably be a field on an activity explicitly mapping it to some item, I would expect?) but regardless, I'm curious why it was closed?
> > > Edit: I've hired somebody to implement this functionality as a custom app for our usecase. When they're done, I'll drop the code here.
> > 
> > 
> > If you're interested, you can add this app to generate itemized line items per timesheet rates on sales invoices https://github.com/508-dev/timesheet_extended/tree/develope
> 
> komali2, your contribution is greatly appreciated . We will endeavor to test it out as soon as possible and give some feedback. Is it currently working well for your use case?

Yes, working quite well, and is FOSS of course so feel free to re-integrate. 

There are probably other features we'll add to it in the future but for now it works for us in generating sales and purchase invoices from timesheets, as well as avoiding generating invoices from timesheets that are already attached to a sales/purchase invoice. 

I think there's a bug in it right now when generating purchase invoices though that I need to look into.
